pubsub/gcloud/aio/pubsub/subscriber_message.py

```diff
@@ -1,75 +1,137 @@
+
+        def __repr__(self) -> str:
+            return repr(self._message)
+
+        @property
+        def attributes(self) -> Any:  # Google .ScalarMapContainer
+            return self._message.attributes
+
+        @property
+        def data(self) -> bytes:
+            return bytes(self._message.data)
+
+        @property
+        def publish_time(self) -> datetime.datetime:
+            published: datetime.datetime = self._message.publish_time
+            return published
+
+        @property
+        def ordering_key(self) -> str:
+            return str(self._message.ordering_key)
+
+        @property
+        def size(self) -> int:
+            return int(self._message.size)
+
+        @property
+        def ack_id(self) -> str:
+            return str(self._message.ack_id)
+
+        @property
+        def delivery_attempt(self) -> Optional[int]:
+            if self._message.delivery_attempt:
+                return int(self._message.delivery_attempt)
+            return None
+
+        def ack(self) -> None:
+            self._message.ack()
+
+        def drop(self) -> None:
+            self._message.drop()
+
+        def modify_ack_deadline(self, seconds: int) -> None:
+            self._message.modify_ack_deadline(seconds)
+
+        def nack(self) -> None:
+            self._message.nack()
+
+else:
+
+    def parse_publish_time(publish_time: str) -> datetime.datetime:
+        try:
+            return datetime.datetime.strptime(
+                publish_time, '%Y-%m-%dT%H:%M:%S.%f%z')
+        except ValueError:
+            return datetime.datetime.strptime(
+                publish_time, '%Y-%m-%dT%H:%M:%S%z')
+
+    class SubscriberMessage:  # type: ignore[no-redef]
+        def __init__(self, ack_id: str, message_id: str,
+                     publish_time: 'datetime.datetime',
+                     data: Dict[str, Any],
+                     attributes: Dict[str, Any]):
+            self.ack_id = ack_id
+            self.message_id = message_id
+            self.publish_time = publish_time
+            self.data = data
+            self.attributes = attributes
+            self._callback = self._default_callback()
+
+        @staticmethod
+        def from_api_dict(received_message: Dict[str, Any]
+                         ) -> 'SubscriberMessage':
+            ack_id = received_message['ackId']
+            message_id = received_message['message']['messageId']
+            data = json.loads(
+                base64.b64decode(received_message['message']['data']))
+            attributes = received_message['message']['attributes']
+            publish_time: datetime.datetime = parse_publish_time(
+                received_message['message']['publishTime'])
+            return SubscriberMessage(ack_id=ack_id, message_id=message_id,
+                                     publish_time=publish_time, data=data,
+                                     attributes=attributes)
+
+        @staticmethod
```

Comment:
I'm not sure these convenience methods are necessary. Do we have any examples of code this will make significantly simpler? Eg. looks like without them acking is:
```python
await subscriber.ack(subscription, [m.ack_id for m in messages])
```
Which seems simple enough that I wouldn't want to complicate things without a good reason.

Reply:
See draft for subscriber implementation code